### ISSUE TITLE
BatchNormalization layer with trainable=False, different behavior in training and testing

### ISSUE THREAD

**Issue by KiraMichiru** [keras-team/keras#9522]
I am not sure whether this i the desired behavior, but I see differences between training and testing mode in the output of BatchNormalization layers of a model that was set to trainable=False. According to the [Keras 2.1.3 release notes](https://github.com/keras-team/keras/releases) I wouldn't have expected this to happen:

> trainable attribute in BatchNormalization now disables the updates of the batch statistics (i.e. if trainable == False the layer will now run 100% in inference mode).

I am running Keras version 2.1.4 and it seems to me that while training, the layer still normalizes using the current mean and variance, rather than the saved moving parameters. Is this possible/desired?

What I am approximately doing (I do not have access to my code at the moment, but I can give an idea:)

`# model is a pre-trained model that uses saved weights and contains some BatchNormalization layers`
`model.trainable=False`
`model.compile(loss='mean_squared_error',optimizer='adam')`
`model.summary() #shows zero trainable weights, as expected`
`expectation = model.predict_on_batch(images)`
`loss1 = model.test_on_batch(images,expectation) # = 0 as expected`
`loss2 = model.train_on_batch(images,expectation) # !=0, contrary to my expectation`
`loss3 = model.test_on_batch(images,expectation) # = 0 as expected`

Edit: Now a fully executable example, tested on both Tensorflow 1.4.0 and 1.6.0:

```
from keras.layers import BatchNormalization
from keras.models import Sequential
from keras.initializers import Constant
import numpy as np

model = Sequential()
bn = BatchNormalization(input_shape=(1,),moving_variance_initializer=Constant(value=0.999)) #modify variance such that moving_variance+epsilon=1
bn.trainable=False
model.add(bn)
model.compile(loss='mean_absolute_error',optimizer='adam')
model.summary() # Total params: 4, Trainable params: 0, Non-trainable params: 4

trial_batch_1 = np.ones(shape=(3,1))    # ones
expectation_1 = np.ones(shape=(3,1))    # without training, default BatchNormalization values should be used, leaving the input unchanged

trial_batch_2 = np.ones(shape=(3,1))
trial_batch_2 = trial_batch_2*2         # twos
expectation_2 = np.ones(shape=(3,1))
expectation_2 = expectation_2*2         # without training, default BatchNormalization values should be used, leaving the input unchanged

print(model.predict_on_batch(trial_batch_1))    # [[ 1.][ 1.][ 1.]]
print(model.predict_on_batch(trial_batch_2))    # [[ 2.][ 2.][ 2.]]

print(model.test_on_batch(trial_batch_1,expectation_1))     # 0.0
print(model.test_on_batch(trial_batch_2,expectation_2))     # 0.0
print(model.train_on_batch(trial_batch_1,expectation_1))    # 1.0
print(model.train_on_batch(trial_batch_2,expectation_2))    # 2.0
print(model.test_on_batch(trial_batch_1,expectation_1))     # 0.0 internal parameters seem to not have changed
print(model.test_on_batch(trial_batch_2,expectation_2))     # 0.0 internal parameters seem to not have changed
```

- [x] If running on TensorFlow, check that you are up-to-date with the latest version. The installation instructions can be found [here](https://www.tensorflow.org/get_started/os_setup).

- [x] Provide a link to a GitHub Gist of a Python script that can reproduce your issue (or just copy the script here if it is short).

**Comment by DNXie** [comment]
I have the same problem.  The `BatchNormalization` in both `keras` and `tensorflow` seems to behave differently in training and testing phase.   Sometimes, it would greatly influence the validation results.

**Comment by twinanda** [comment]
Same here. I load the same model weights to two different models, one with training=True for the BN parameter and another one with training=False. The one with training=True gives reasonable results, while training=False gives very bad result.

**Comment by fferroni** [comment]
Same here

**Comment by cipher009** [comment]
Same here. Any suggestions about this?

**Comment by EduCM** [comment]
http://blog.datumbox.com/the-batch-normalization-layer-of-keras-is-broken/

**Comment by Evan05543071** [comment]
https://stackoverflow.com/questions/50047653/set-training-false-of-tf-layers-batch-normalization-when-training-will-get-a

**Comment by alecGraves** [comment]
As mentioned in https://github.com/keras-team/keras/pull/9965#issuecomment-382801648, you want to put the layer in inference mode.

`layer.trainable` is meant to only affect weight updates, not put the layer in inference mode.

You can put the layer in inference mode by calling the bn layer with `training=False` when using the functional API, or in your case:

```python
import keras.backend as K
K.set_learning_phase(0)
# ... your code here
```

This will put the layer in inference mode. Optionally, you can put the model in inference mode after you have already instantiated it by running:
```python
# ... your code here
print(model.test_on_batch(trial_batch_1,expectation_1))     # 0.0
print(model.test_on_batch(trial_batch_2,expectation_2))     # 0.0
print(model.train_on_batch(trial_batch_1,expectation_1))    # 1.0
print(model.train_on_batch(trial_batch_2,expectation_2))    # 2.0
print(model.test_on_batch(trial_batch_1,expectation_1))     # 0.0 internal parameters seem to not have changed
print(model.test_on_batch(trial_batch_2,expectation_2))     # 0.0 internal parameters seem to not have changed

import keras.backend as K
K.set_learning_phase(0)
config = model.get_config()
weights = model.get_weights()
model = Sequential.from_config(config)
model.set_weights(weights)
model.compile(loss='mean_absolute_error',optimizer='adam')

print(model.test_on_batch(trial_batch_1,expectation_1))     # 0.0
print(model.test_on_batch(trial_batch_2,expectation_2))     # 0.0
print(model.train_on_batch(trial_batch_1,expectation_1))    # 0.0 no change
print(model.train_on_batch(trial_batch_2,expectation_2))    # 0.0 no change
print(model.test_on_batch(trial_batch_1,expectation_1))     # 0.0
print(model.test_on_batch(trial_batch_2,expectation_2))     # 0.0

```

**Comment by moondra2017** [comment]
> As mentioned in [#9965 (comment)](https://github.com/keras-team/keras/pull/9965#issuecomment-382801648), the layer must manually be placed in inference mode to keep constant mean and variance during training.
> 
> `layer.trainable` is meant to only affect weight updates, not put the layer in inference mode.
> 
> The patch notes seem to not match the real world code.
> 
> The layer can be placed in inference mode by calling with `training=False` when using the functional API, or in this case:
> 
> ```python
> import keras.backend as K
> K.set_learning_phase(0)
> # ... your code here
> print(model.test_on_batch(trial_batch_1,expectation_1))     # 0.0
> print(model.test_on_batch(trial_batch_2,expectation_2))     # 0.0
> print(model.train_on_batch(trial_batch_1,expectation_1))    # 0.0 no change
> print(model.train_on_batch(trial_batch_2,expectation_2))    # 0.0 no change
> print(model.test_on_batch(trial_batch_1,expectation_1))     # 0.0
> print(model.test_on_batch(trial_batch_2,expectation_2))     # 0.0
> ```
> 
> As seen from the output, this put the layer in inference mode. Optionally, inference mode can be set after the model has been loaded with:
> 
> ```python
> # ... your code here
> # default dynamic mode:
> print(model.test_on_batch(trial_batch_1,expectation_1))     # 0.0
> print(model.test_on_batch(trial_batch_2,expectation_2))     # 0.0
> print(model.train_on_batch(trial_batch_1,expectation_1))    # 1.0
> print(model.train_on_batch(trial_batch_2,expectation_2))    # 2.0
> print(model.test_on_batch(trial_batch_1,expectation_1))     # 0.0 internal parameters seem to not have changed
> print(model.test_on_batch(trial_batch_2,expectation_2))     # 0.0 internal parameters seem to not have changed
> 
> import keras.backend as K
> K.set_learning_phase(0) # inference mode
> config = model.get_config()
> weights = model.get_weights()
> model = Sequential.from_config(config)
> model.set_weights(weights)
> model.compile(loss='mean_absolute_error',optimizer='adam')
> 
> print(model.test_on_batch(trial_batch_1,expectation_1))     # 0.0
> print(model.test_on_batch(trial_batch_2,expectation_2))     # 0.0
> print(model.train_on_batch(trial_batch_1,expectation_1))    # 0.0 no change
> print(model.train_on_batch(trial_batch_2,expectation_2))    # 0.0 no change
> print(model.test_on_batch(trial_batch_1,expectation_1))     # 0.0
> print(model.test_on_batch(trial_batch_2,expectation_2))     # 0.0
> 
> K.set_learning_phase(1) # training mode
> weights = model.get_weights()
> model = Sequential.from_config(config)
> model.set_weights(weights)
> model.compile(loss='mean_absolute_error',optimizer='adam')
> 
> print(model.test_on_batch(trial_batch_1,expectation_1))     # 1.0
> print(model.test_on_batch(trial_batch_2,expectation_2))     # 2.0
> print(model.train_on_batch(trial_batch_1,expectation_1))    # 1.0
> print(model.train_on_batch(trial_batch_2,expectation_2))    # 2.0
> print(model.test_on_batch(trial_batch_1,expectation_1))     # 1.0
> print(model.test_on_batch(trial_batch_2,expectation_2))     # 2.0
> ```

I think setting `training = False` doesn't seem to keep the frozen batch_norm in inference mode:

https://github.com/keras-team/keras/issues/12400

I just started following this issue, as I have also noticed some inconsistency with my models and haven't read through all of the threads, so still trying to catch up.

In addition, if I set learning_phase to 0 during training, wouldn't that mean `dropout` wouldn't work?

**Comment by Hayashi-Yudai** [comment]
keras-team/keras#7085 (comment)
This will help you.
